### ISSUE TITLE
deadlock in function constructTime with timerType = TTYPE_YEARLY_WD

### ISSUE THREAD

**Issue by gizmocuz** [domoticz/domoticz#1194]
There is a deadlock situation in the 'constructTime' function when this is called when the timerType equals TTYPE_YEARLY_WD

Scheduler.cpp, line 545

`else if (pItem->timerType == TTYPE_YEARLY_WD)`

@gordonb3, is it possible you can have a peak into this one ?
![image](https://cloud.githubusercontent.com/assets/2265299/22562903/c43ef9e4-e97e-11e6-9c22-b563ef9d2938.png)

If you add a dummy switch and add a timer like this, then you will experience this.

Please note, i prevented this in the current code (loop counter), but thats not how to solve this i suppose, i dont know why this is happening

Thanks!

**Comment by gordonb3** [comment]
I had a quick look but don't see any obvious error. A possible issue in this case might be that in CET there is no 2:00 AM on March Sunday last, but I'd think mktime() should return a valid time of 3:00 AM CETD in this case.

I'll need to set a trace to confirm what is happening. One question: is this problem specific to any OS?

**Comment by gizmocuz** [comment]
The original user with the problem was using Domoticz on a raspberry, and I have tested/debugged this under windows.
Now... this is a strange date/time of course... it's the moment when we switch between summer/winter time, maybe that has something to do with it...

**Comment by gordonb3** [comment]
Yes I did notice that . It's in my previous comment. The thing here is that this particular request to constructTime() does not set a specific DST but asks the time functions to set one that corresponds with the given time.

Because DST is set to "request" (-1) the function should not loop at all. There is however a check that is performed earlier, which is meant to address a call where DST is set (1) but the current time zone setting says there is no DST.

This would mean that this particular time creates some type of loop where the time functions return result "invalid" (time = -1) but does set the DST flag in the corresponding time struct. Again I need to verify the who what and why.

Of course if the suspicion turns out to be true, solving the loop will only be half of the answer. Because the function returns time invalid the corresponding schedule will never fire.

**Comment by gordonb3** [comment]
Okay. I created a simple test application feeding the values for March 26 and October 29 directly into the constructTime() function and I cannot replicate the issue. Which must mean that these are in fact not the values loaded by the TTYPE_YEARLY_WD scheduler procedure.

Will try to find out what these values are later.

**Comment by gizmocuz** [comment]
Best to add a timer like above, and restart domoticz
if you set a breakpoint (in visual studio) in the (pItem->timerType == TTYPE_YEARLY_WD) function, you should experience the same...
But... do remember i made a small patch that prevents the endless loop in the 'constructTime' function

**Comment by gordonb3** [comment]
Got it!

You actually made two changes. The original code does not ask for the DST state but assumes the state taken from the current time, which is common throughout the code so the issue is not restricted to this particular timer type. When I added that to my test application the issue popped up immediately. What happens here is that for this particular hour (March Sunday last 2:00 AM - 2:59 AM) the DST flag will always toggle regardless of what value you feed it because the input always specifies an incorrect time.

I'll have a definitive fix shortly.

**Comment by gizmocuz** [comment]
Sorry yes.... forgot (to mention) this ;)
Great... i already thought there was something different with this date... maybe also for the October date ?

**Comment by gordonb3** [comment]
Nothing wrong with the October date. The only issue there is that both DST states are valid for times between 2:00AM and 3:00AM. Effectively this creates an uncertainty at what precise time the event will trigger but that really can't be helped.

There does appear to be some inconsistency between platforms though. I did my earlier test on Linux and the 2:00 AM time translated into 3:00 AM with DST set. When I fired up the Windows dev machine and ran the check there it returned 1:00 AM with DST off. I don't think that's the result we want to see here.
